auslib/web/admin/emergency_shutoff.py

```diff
@@ -0,0 +1,104 @@
+import json
+import time
+from flask import Response
+from connexion import problem
+from auslib.global_state import dbo
+from auslib.web.admin.views.base import (
+    requirelogin, handleGeneralExceptions, transactionHandler)
+from auslib.web.admin.views.scheduled_changes import (
+    EnactScheduledChangeView, ScheduledChangesView, SignoffsView)
+
+
+def get_emergency_shutoff(product, channel):
+    where = dict(product=product, channel=channel)
+    shutoffs = dbo.emergencyShutoffs.select(where=where)
+    return shutoffs[0] if shutoffs else None
+
+
+def shutoff_already_exists(product, channel):
+    return bool(get_emergency_shutoff(product, channel))
+
+
+def product_requires_signoffs(product, channel):
+    rs = dbo.productRequiredSignoffs.select(where=dict(product=product, channel=channel))
+    return bool(rs)
+
+
+def process_shutoff_post(emergency_shutoff):
+    if 'csrf_token' in emergency_shutoff:
+        del emergency_shutoff['csrf_token']
+    if 'data_version' not in emergency_shutoff:
+        emergency_shutoff['data_version'] = 1
+    return emergency_shutoff
+
+
+@requirelogin
+@transactionHandler
+@handleGeneralExceptions('POST')
+def post(emergency_shutoff, changed_by, transaction):
+    if shutoff_already_exists(emergency_shutoff['product'], emergency_shutoff['channel']):
+        return problem(
+            400, 'Bad Request', 'Invalid Emergency shutoff data',
+            ext={'data': 'Emergency shutoff for product/channel already exists.'})
+    if not product_requires_signoffs(emergency_shutoff['product'], emergency_shutoff['channel']):
+        return problem(
+            400, 'Bad Request', 'Invalid Emergency shutoff data',
+            ext={'data': 'The given product/channel should requires signoffs.'})
+    emergency_shutoff = process_shutoff_post(emergency_shutoff)
+    dbo.emergencyShutoffs.insert(
+        changed_by=changed_by, transaction=transaction, **emergency_shutoff)
+    schedule_reenable_updates(emergency_shutoff, changed_by, transaction)
```

Comment:
Hm, this is interesting! I wasn't expecting this to be scheduled right at shutoff time - I would've expected a user to need to hit a button to schedule it. I guess this is why you disabled auto signoff for the new table? I'm curious why you chose to do it here instead of in a separate endpoint.

Reply:
This is to achieve two requirements: 
1) anyone with permission can disable updates;
2) only experienced (releng, relman, etc) people can reenable updates.

So the way that I thought to implement that was:
- User disables the updates by product/channels;
- At the same moment of deactivation, an ASAP change is scheduled;
- When all is OK with product/channel, it is necessary only  perform signoff to reenable the updates. This behaviour will work only if the product requires signoffs.

Reply:
Two things stand out to me about this:
1) Can it be made to work for product+channels that don't require signoffs (like Firefox nightly) - that's a hard requirement for feature.
2) How will this be presented in the UI? I'm a bit concerned that people may see the re-enabling waiting on signoff, and assume that it's ready to be turned back on.

Reply:
Hmmm that makes sense.
So, we can simplify more:
-Who has the 'create' permission for emergency shutoff, can disable updates.
-Who has the 'delete' permission for emergency shutoff, can reenable updates.

No need to schedule changes nor ask to sign offs. What do you think?

Reply:
For product+channels that don't have signoffs defined - yes (eg: Firefox nightly). For product+channels that do have signoffs defined, re-enabling updates must bgo through multiple signoffs still.